### PR TITLE
[BEAM-5112] Use Calcite codegen to implement BeamCalcRel

### PR DESCRIPTION
**Pull request by apilloud**
This replaces our SQL function implementation with Calcite's. Passes all tests when run against calcite master with the following PRs:

 - https://github.com/apache/calcite/pull/818
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/planner/BeamRelDataTypeSystem.java [hunk 1]
```diff
@@ -35,4 +35,9 @@ public int getMaxNumericScale() {
   public int getMaxNumericPrecision() {
     return 38;
   }
+
+  @Override
+  public boolean shouldConvertRaggedUnionTypesToVarying() {
+    return true;
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/planner/BeamRelDataTypeSystem.java]: nit: we should add comments about why all the stuff in this class is set to these values

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 2]
```diff
@@ -68,15 +119,128 @@ public Calc copy(RelTraitSet traitSet, RelNode input, RexProgram program) {
...
+      List<Expression> expressions =
+          RexToLixTranslator.translateProjects(
+              program,
+              typeFactory,
+              builder,
+              physType,
+              DataContext.ROOT,
+              new InputGetterImpl(input, upstream.getSchema()),
+              null);
+      assert expressions.size() == outputSchema.getFieldCount();
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: should this be an exception?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 3]
```diff
@@ -68,15 +119,128 @@ public Calc copy(RelTraitSet traitSet, RelNode input, RexProgram program) {
...
+
+      // output = Row.withSchema(outputSchema)
+      Expression output = Expressions.call(Row.class, "withSchema", outputSchemaParam);
+      Method addValue = Types.lookupMethod(Row.Builder.class, "addValue", Object.class);
+
+      for (int index = 0; index < expressions.size(); index++) {
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: This should be split into methods, otherwise it's hard to follow

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 4]
```diff
@@ -58,8 +98,19 @@ public Calc copy(RelTraitSet traitSet, RelNode input, RexProgram program) {
...
+          .put(TypeName.INT16, Short.class)
+          .put(TypeName.INT32, Integer.class)
+          .put(TypeName.INT64, Long.class)
+          .put(TypeName.FLOAT, Float.class)
+          .put(TypeName.DOUBLE, Double.class)
+          .build();
+
   private class Transform extends PTransform<PCollectionList<Row>, PCollection<Row>> {
 
+    /** expand is based on calcite's EnumerableCalc.implement(). */
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Can you add a description of how stuff works here?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/planner/BeamRelDataTypeSystem.java (hunk 1 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/planner/BeamRelDataTypeSystem.java]: Added 'operators that change string length should return varchar.'

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java (hunk 2 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: It is something that should never be false. I will just delete it.

**Comment by apilloud** [comment]
run java postcommit

**Comment by apilloud** [comment]
Run Dataflow Runner Nexmark Tests

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/integrationtest/BeamSqlDateFunctionsIntegrationTest.java [hunk 5]
```diff
@@ -55,11 +56,13 @@ public Void apply(Iterable<Row> input) {
...
-      assertTrue(date.getMillis() - row.getDateTime(0).getMillis() < 1000);
-      assertTrue(date.getMillis() - row.getDateTime(1).getMillis() < 1000);
-      assertTrue(date.getMillis() - row.getDateTime(2).getMillis() < 1000);
-      assertTrue(date.getMillis() - row.getDateTime(3).getMillis() < 1000);
-      assertTrue(date.getMillis() - row.getDateTime(4).getMillis() < 1000);
+      long millis = date.getMillis();
+      int timeMillis = (int) (date.getMillis() % MILLIS_PER_DAY);
+      assertTrue(timeMillis - row.getDateTime(0).getMillis() < 1000);
+      assertTrue(millis - row.getDateTime(1).getMillis() < 1000);
+      assertTrue(millis - row.getDateTime(2).getMillis() < (MILLIS_PER_DAY + 1000));
```
Comment (akedin) [inline sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/integrationtest/BeamSqlDateFunctionsIntegrationTest.java]: nit: can you add few comments here, it's hard to parse these assertions

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlDslSqlStdOperatorsTest.java [hunk 6]
```diff
@@ -1047,7 +1046,9 @@ public void testTruncate() {
   @SqlOperatorTest(name = "RAND", kind = "OTHER_FUNCTION")
   public void testRand() {
     ExpressionChecker checker =
-        new ExpressionChecker().addExpr("RAND(c_integer)", new Random(INTEGER_VALUE).nextDouble());
+        new ExpressionChecker()
+            .addExpr(
+                "RAND(c_integer)", new Random(INTEGER_VALUE ^ (INTEGER_VALUE << 16)).nextDouble());
```
Comment (akedin) [inline sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlDslSqlStdOperatorsTest.java]: So Beam SQL rand is always initialized with the same seed? Is this what we want?

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlCastTest.java [hunk 7]
```diff
@@ -67,35 +68,7 @@ public void testCastToDate() {
 
     PAssert.that(result)
         .containsInAnyOrder(
-            Row.withSchema(resultType).addValues(1, new DateTime(2018, 10, 18, 0, 0)).build());
-
-    pipeline.run();
-  }
-
-  @Test
-  public void testCastToDate2() {
```
Comment (akedin) [inline sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlCastTest.java]: What's wrong with this one?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 8]
```diff
@@ -68,15 +121,90 @@ public Calc copy(RelTraitSet traitSet, RelNode input, RexProgram program) {
           BeamCalcRel.class.getSimpleName(),
           pinput);
       PCollection<Row> upstream = pinput.get(0);
+      Schema outputSchema = CalciteUtils.toSchema(getRowType());
+
+      final SqlConformance conformance = SqlConformanceEnum.MYSQL_5;
+      final JavaTypeFactory typeFactory =
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: I think `conformance` and typeFactory should be moved out into a separate class similar to `BeamRelDataTypeSystem`

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 9]
```diff
@@ -68,15 +121,90 @@ public Calc copy(RelTraitSet traitSet, RelNode input, RexProgram program) {
...
+
+      final RexBuilder rexBuilder = getCluster().getRexBuilder();
+      final RelMetadataQuery mq = RelMetadataQuery.instance();
+      final RelOptPredicateList predicates = mq.getPulledUpPredicates(getInput());
+      final RexSimplify simplify = new RexSimplify(rexBuilder, predicates, false, RexUtil.EXECUTOR);
+      final RexProgram program = BeamCalcRel.this.program.normalize(rexBuilder, simplify);
 
-      BeamSqlExpressionExecutor executor = new BeamSqlFnExecutor(BeamCalcRel.this.getProgram());
+      Expression condition =
+          RexToLixTranslator.translateCondition(
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Should there be some validation of what is actually in `program` or is it always safe to call `translateCondition`? What does it do?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 10]
```diff
@@ -68,15 +121,90 @@ public Calc copy(RelTraitSet traitSet, RelNode input, RexProgram program) {
...
+        output = Expressions.call(output, addValue, castOutput(value, toType));
+      }
+
+      // .build();
+      output = Expressions.call(output, "build");
+
+      // if (condition) {
+      //   c.output(output);
+      // }
+      builder.add(
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Some explanation of this and an example of the generated code would help here

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 11]
```diff
@@ -96,43 +224,243 @@ public boolean isInputSortRelAndLimitOnly() {
...
+      ImmutableMap.<TypeName, Type>builder()
+          .put(TypeName.BYTE, Byte.class)
+          .put(TypeName.INT16, Short.class)
+          .put(TypeName.INT32, Integer.class)
+          .put(TypeName.INT64, Long.class)
+          .put(TypeName.FLOAT, Float.class)
+          .put(TypeName.DOUBLE, Double.class)
+          .build();
+
+  private Expression castOutput(Expression value, FieldType toType) {
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: This probably belongs to `CalciteUtils`, we have similar casting logic there

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 12]
```diff
@@ -96,43 +224,243 @@ public boolean isInputSortRelAndLimitOnly() {
...
+    if (value.getType() == Object.class) {
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Is this possible to get an expression with `Object` type? Would it be safer to throw here to avoid breaking something in our schema logic?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 13]
```diff
@@ -96,43 +224,243 @@ public boolean isInputSortRelAndLimitOnly() {
...
+      // fast copy path, just pass object through
+      return value;
+    } else if (toType.getTypeName().isDateType()
+        && value.getType() instanceof Class
+        && !Types.isAssignableFrom(ReadableInstant.class, (Class) value.getType())) {
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Isn't `instanceof Class` and cast to `Class` redundant here?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 14]
```diff
@@ -96,43 +224,243 @@ public boolean isInputSortRelAndLimitOnly() {
...
+      Expression valueDateTime = value;
+      if (toType.getMetadata() == null) {
+        if (value.getType() == java.sql.Timestamp.class) {
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Else throw in these ifs?
Do you need to compare metadata? Why isn't it enough to compare `CalciteUtils.TIME.equals(toType)`? Would it be easier to read in `switch/case` format?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 15]
```diff
@@ -96,43 +224,243 @@ public boolean isInputSortRelAndLimitOnly() {
...
+          .put(TypeName.INT64, Long.class)
+          .put(TypeName.FLOAT, Float.class)
+          .put(TypeName.DOUBLE, Double.class)
+          .build();
+
+  private Expression castOutput(Expression value, FieldType toType) {
+    if (value.getType() == Object.class) {
+      // fast copy path, just pass object through
+      return value;
+    } else if (toType.getTypeName().isDateType()
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Can you comment the datetime logic a bit? Maybe extract into a separate method?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java [hunk 16]
```diff
@@ -96,43 +224,243 @@ public boolean isInputSortRelAndLimitOnly() {
...
+      return null;
+    }
+
+    @Override
+    public QueryProvider getQueryProvider() {
+      return null;
+    }
+
+    @Override
+    public Object get(String name) {
```
Comment (akedin) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Add a comment about what this is?

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlDslSqlStdOperatorsTest.java (hunk 6 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlDslSqlStdOperatorsTest.java]: This is actually testing Calcite now. I'm guessing this behavior is determined by the SQL standard, but we probably don't do the right thing if there is more than one thread/worker.

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlCastTest.java (hunk 7 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamSqlCastTest.java]: 20181018 means year 20181018. This results in a overflow.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java (hunk 9 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: It should be always safe to call. It validates the program has a condition and if not emits 'true'. The code generator will optimize away 'if (true)' before generating the java code.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java (hunk 10 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: An example of the generated code is in the comment immediately above this block.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java (hunk 11 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: It is never used outside of this class, I think leaving it here makes this code easier to understand. If we have a second use case for it we can move it out to `CalciteUtils`.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java (hunk 12 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: See the comment below. This is the fast copy path, which is actually used quite a bit. When a field is copied from input to output without modification code is generated to fetch it with `getValue` which returns an object. That object can be passed through to the new row without any conversion.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java (hunk 16 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Added. `/* DataContext.get is used to fetch "global" state inside the generated code */`

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java (hunk 13 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: No, it isn't. There are things that don't resolve to a type that come through here. I believe constants are one example. I will change this so they exit in the fast path too.

**Comment by akedin** [comment]
And what's the perf change after this?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java (hunk 14 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: The code generator only converts to the `java.sql` types if there are UDFs, otherwise they will remain primitive types. So we don't always need to convert back.
I am comparing metadata because `equals` compares things I don't care about here, like nullability. Looks like there is a new `typesEqual` helper I can use here.
I don't think this logic is compatible with a `switch/case` format.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java (hunk 15 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java]: Both done.

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/integrationtest/BeamSqlDateFunctionsIntegrationTest.java (hunk 5 above)
Comment (apilloud) [reply sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/integrationtest/BeamSqlDateFunctionsIntegrationTest.java]: This is a bad test. I've cleaned it up significantly.

**Comment by apilloud** [comment]
This is a 12% speedup in a 20 second run and 22% in 60 second run of Nexmark Query 1. It was a significant speedup on a ~30 minute run, but it doesn't look like we have one of those in our postcommit anymore. Jobs that are very short or lite on Calc logic might see a slowdown. There is probably some opportunity for optimization still.

**Comment by kanterov** [comment]
This PR could be possibly the root cause of https://issues.apache.org/jira/browse/BEAM-6401
